ComponentKit/DataSources/Common/CKComponentAnnouncerHelper.h

```diff
@@ -0,0 +1,59 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <objc/message.h>
+#import <vector>
+
+#import <Foundation/Foundation.h>
+#import "CKComponentAnnouncerBase.h"
+#import "CKComponentAnnouncerBaseInternal.h"
+
+namespace CK {
+  namespace Component {
+    struct AnnouncerHelper
+    {
+    private:
+      // function to load the current listeners vector in a thread safe way
+      static std::shared_ptr<const std::vector<__weak id>> loadListeners(CKComponentAnnouncerBase *self);
+    public:
+      template<typename... ARGS>
+      static void call(CKComponentAnnouncerBase *self, SEL s, ARGS... args) {
+        typedef void (*TT)(id self, SEL _cmd, ARGS...); // for floats, etc, we need to use the strong typed versions
+        TT objc_msgSendTyped = (TT)(void*)objc_msgSend;
+        
+        auto frozenListeners = loadListeners(self);
+        if (frozenListeners) {
+          for (id listener : *frozenListeners) {
+            objc_msgSendTyped(listener, s, args...);
+          }
+        }
+      }
+      
+      template<typename... ARGS>
+      static void callOptional(CKComponentAnnouncerBase *self, SEL s, ARGS... args) {
```

Comment:
1. Not sure, digging into history didn't reveal any discussion around this. I assume it was done in order to make announcing as efficient as possible (cpu and memory wise).
2. Unfortunately not, because templates. I'm not an expert in C++ either, but I found this explanation: http://stackoverflow.com/questions/495021/why-can-templates-only-be-implemented-in-the-header-file